subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java

```diff
@@ -190,24 +119,98 @@ public void setTasks(Collection<String> tasks) {
      * @since 6.0
      */
     @Internal
-    @ToBeReplacedByLazyProperty
-    public String getBuildName() {
-        return buildName;
-    }
-
-    /**
-     * Sets build name to use for the nested build.
-     *
-     * @param buildName the build name to use for the nested build
-     * @since 6.0
-     */
-    public void setBuildName(String buildName) {
-        this.buildName = buildName;
-    }
+    @Optional
+    @ReplacesEagerProperty
+    public abstract Property<String> getBuildName();
 
     @TaskAction
     void build() {
         // TODO: Allow us to inject plugins into nested builds too.
-        runNestedRootBuild(buildName, (StartParameterInternal) getStartParameter(), getServices());
+        runNestedRootBuild(getBuildName().getOrNull(), (StartParameterInternal) getStartParameter(), getServices());
+    }
+
+    static class StartParameterAdapter {
+        /**
+         * Returns the full set of parameters that will be used to execute the build.
+         *
+         * @return the parameters. Never returns null.
+         */
+        @BytecodeUpgrade
+        static StartParameter getStartParameter(GradleBuild task) {
+            return task.getStartParameter().get();
+        }
+
+        /**
+         * Sets the full set of parameters that will be used to execute the build.
+         *
+         * @param startParameter the parameters. Should not be null.
+         */
+        @BytecodeUpgrade
+        static void setStartParameter(GradleBuild task, StartParameter startParameter) {
+            task.getStartParameter().set(startParameter);
+        }
+
+        /**
+         * Returns the project directory for the build. Defaults to the project directory.
+         *
+         * @return The project directory. Never returns null.
+         */
+        @Internal
+        @BytecodeUpgrade
+        static File getDir(GradleBuild task) {
+            return task.getStartParameter().get().getCurrentDir();
+        }
+
+        /**
+         * Sets the project directory for the build.
+         *
+         * @param dir The project directory. Should not be null.
+         * @since 4.0
+         */
+        @BytecodeUpgrade
+        static void setDir(GradleBuild task, File dir) {
+            setDir(task, (Object) dir);
+        }
+
+        /**
+         * Sets the project directory for the build.
+         *
+         * @param dir The project directory. Should not be null.
+         */
+        @BytecodeUpgrade
+        static void setDir(GradleBuild task, Object dir) {
+            task.getStartParameter().get().setCurrentDir(task.getProject().file(dir));
```

Comment:
🤔 I wonder if it's a good pattern. 

We need to mutate the property. Here, we resolve it and mutate.

What we probably need is to add a transformation that applies lazily. Something like `getStartParameter().transform(x -> x.someAction())`. 
A workaround: 
```
getStartParameter().set(getStartParameter().map(x -> someFunc(x)))
```
But I don't like it and don't think it's idiomatic. Another complication here is that `setCurrentDir` returns void, so it should be `map(x -> { someFunc(x); return x;} )`

Reply:
I was also thinking about flattening `startParameter`, exposing a property per it's field, having read-only `getStartParameter` to be constructed on the fly and `setStartParameter` delegating to properties. But there are too many properties for doing that, and the process will be clunky too.
 
 I'll create an issue for that.

Reply:
The related issue:
* https://github.com/gradle/gradle/issues/27227